src/fal/planner/plan.py

```diff
@@ -41,16 +67,73 @@ def from_execution_plan(
 
 
 @dataclass
-class PlannedGraph(OriginGraph):
+class ShuffledGraph(OriginGraph):
     graph: nx.DiGraph
+    after_pattern = re.compile("^script\\..+\\.AFTER\\..+\\.(py|ipynb)")
+    before_pattern = re.compile("^script\\..+\\.BEFORE\\..+\\.(py|ipynb)")
 
     @classmethod
     def from_filtered_graph(
         cls,
         filtered_graph: FilteredGraph,
-        enable_chunking: bool = True,
     ):
         graph = filtered_graph.copy_graph()
+        shuffled_graph = cls(graph)
+        shuffled_graph._shuffle()
+        return shuffled_graph
+
+    def _shuffle(self):
+        def _pattern_matching(pattern: re.Pattern, nodes: Set[str]):
+            matched = {
+                maybe_script for maybe_script in nodes if pattern.match(maybe_script)
+            }
+            return matched, nodes - matched
+
+        def _add_edges_from_to(from_nodes: Set[str], to_nodes: Set[str]):
+            self.graph.add_edges_from(
+                (from_n, to_n) for from_n in from_nodes for to_n in to_nodes
+            )
+
+        old_graph = self.copy_graph()
+        node: str
+        for node in old_graph.nodes:
+
+            # TODO: is_after_script() later on!!
+            after_scripts, other_succs = _pattern_matching(
+                ShuffledGraph.after_pattern, set(old_graph.successors(node))
+            )
+            # Keep the original node to succs edges and add a new one from the script to succs
+            _add_edges_from_to(after_scripts, other_succs)
```

Comment:
You can also remove the edges from `node` to `other_succs` (for better calculation in scheduler's dependency management code)

Reply:
Thinking...

```mermaid
flowchart LR
  a --> b & a_after1 & a_after2;
```

with your proposal, it becomes

```mermaid
flowchart LR
  a --> a_after1 & a_after2;
  a_after1 & a_after2 --> b;
```

current solution does

```mermaid
flowchart LR
  a --> a_after1 & a_after2 & b;
  a_after1 & a_after2 --> b;
```

*****

Regarding selection:

`--select a+1`: 

```mermaid
flowchart LR
  a --> a_after1 & a_after2 & b;
  a_after1 & a_after2 --> b;
  classDef running fill:green;
  class a,a_after1,a_after2,b running;
```

vs


```mermaid
flowchart LR
  a --> a_after1 & a_after2;
  a_after1 & a_after2 --> b;
  classDef running fill:green;
  class a,a_after1,a_after2 running;
```

Reply:
But doesn't the selection happen before we re-order the graph @chamini2? We calculate the execution plan according to the *original graph* (without re-ordering) and then once we get the `FilteredGraph` we can safely re-order everything without worrying about how it would change the behavior for the users.